### ISSUE TITLE
Log disappears if network not connected

### ISSUE THREAD

**Issue by VedantParanjape** [VedantParanjape/esp-wifi-logger#9]
If network is not yet connected, then send function will still pop the queue and fail to do so, but if in future network did connect the specific message will be lost.

@shivam5594 I gave this problem some thought. I came up with 3 solutions:

1) make the function `tcp_network_manager()` blocking, i.e. make it retry to connect until it does like this

```c
int err = -1;
while(err != 0)
{
   err = connect(nm->sock, (struct sockaddr *)&nm->dest_addr, sizeof(nm->dest_addr));
}
    if (err != 0) {
        ESP_LOGE(TAG, "Socket unable to connect: errno %d", errno);
        tcp_close_network_manager(nm);
        nm->sock = -1;
    }
    else
    {
        ESP_LOGI(TAG, "%s", "Successfully connected");
    }
```

2) Use `xQueuePeek()` to receive the front element in the message queue, and if it sent successfully, only then pop it from queue using `xQueueReceive()`, changes will be needed in send function. Reference: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html#_CPPv410xQueuePeek13QueueHandle_tPCv10TickType_t

3) Make the send function blocking by adding something like this at the top, so that the function won't go into popping and receiving part:
`while(is_network_connected());`
But, then in this case we need to implement a event handler which will try to connect to the server again and again, just like the websocket handler does.

In my opinion, the 3rd option is going to be the most appropriate and neat !

**Comment by shivam5594** [comment]
Sorry commented there, missed this. [My answer](https://github.com/VedantParanjape/esp-wifi-logger/issues/7#issuecomment-722183012)

**Comment by VedantParanjape** [comment]
> > Okay, But I am unable to understand how will a semaphore solve this issue, can you elaborate more on this.
> 
> If we wait until the ESP driver changes the state to connected, and once it is changed to that, we know it is connected so releasing the semaphore will make sure that even the efforts are reading from queue and sending will be done after the successful connection.
> 
> Additionally while(); won't give up the CPU since it is a blocking function. If we wait for a semaphore in while() inside main(), it will make sure that other tasks get the CPU. For us we get all the traces made even before connection was established.

Okay this makes sense. As in freertos will pause the function until network connects. But it will not block the CPU, rather run some other process in this time being, right ? I thought even with simple while() freertos will do exactly the same.

> > I don't think I can do that, freeRTOS port of esp32 hasn't documented that function, even though it is there in the header file.
> > But what are the tradeoff of using QueuePeek vs QueueSendToFront
> 
> I see that function in ESP32 FreeRTOS [doc](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html#c.xQueueSendToFront).
> I was just suggesting in case you don't want to change other functions like xQueueReceive().
> 
> QueuePeek: is a better idea I feel. Might have to change few more lines.
> QueueSendToFront: Quick fix and might not have to change the current code flow.

Oh, it was down below in macro section, missed it. I'll go with QueuePeek solution.

@shivam5594 Would you like to solve this issue or any subparts of it let me know, I'll start working on the remaining stuff.

**Comment by shivam5594** [comment]
>>>Okay, But I am unable to understand how will a semaphore solve this issue, can you elaborate more on this.

>>If we wait until the ESP driver changes the state to connected, and once it is changed to that, we know it is connected so releasing the semaphore will make sure that even the efforts are reading from queue and sending will be done after the successful connection.

>>Additionally while(); won't give up the CPU since it is a blocking function. If we wait for a semaphore in while() inside main(), it will make sure that other tasks get the CPU. For us we get all the traces made even before connection was established.

>Okay this makes sense. As in freertos will pause the function until network connects. But it will not block the CPU, rather run some other process in this time being, right ? I thought even with simple while() freertos will do exactly the same.

Yes, you are right. Basically all FreeRTOS tasks are while loops. They don't give up CPU until we use vTaskDelay(), vTaskDelayUntil() or some thread safe blocking functions like wait on semaphore.

>@shivam5594 Would you like to solve this issue or any subparts of it let me know, I'll start working on the remaining stuff.

Already started for websocket for now. :wink: :smiley:

**Comment by VedantParanjape** [comment]
Cool 👍🏼

**Comment by VedantParanjape** [comment]
Hi @shivam5594 any updates ?

**Comment by shivam5594** [comment]
@VedantParanjape Websocket is done without using semaphore. 
- Will wait until port is up 
- No data sending or Queue popping until connection is up
- If it gets disconnected in the middle of it, it will wait again until it comes up (until queue is not full otherwise gives error) and then pushes all the previous messages since the connection was dropped

working on TCP. Will send PR soon.

**Comment by VedantParanjape** [comment]
@shivam5594 This will blowup the queue with error messages quite soon. Can you do something like if the state remains same then same error message won't get printed. This is for websocket handler.

**Comment by shivam5594** [comment]
Yes already done that. I have made required changes. 👍🏼

**Comment by VedantParanjape** [comment]
Sounds good

**Comment by shivam5594** [comment]
PR [created](https://github.com/VedantParanjape/esp-wifi-logger/pull/11).

**Comment by VedantParanjape** [comment]
I'll review this and merge it, give me a few days, submissions going on

**Comment by shivam5594** [comment]
It is useless for UPD connection since errno is always 0 for it, whether server is up or not. Not sure how to check if it is connected. Leaving it as it is.

**Comment by VedantParanjape** [comment]
Sure, UDP is a connection less. So it is not necessary anyways.

Connection less as in it doesn't care if message was received, it simple broadcasts.

**Comment by VedantParanjape** [comment]
Fixed in #11
